src/interface/swarm_comms.cpp

```diff
@@ -236,42 +240,102 @@ void Swarm::LoadBuffers_() {
-      KOKKOS_LAMBDA(const int n) { buffer_counters[n] = 0; });
+  if(max_active_index_ >= 0) {
+    // Make an n particle sized array of index, buffer pairs (with SwarmKey struct)
+    ParArray1D<SwarmKey> buffer_sorted("buffer_sorted", max_active_index_+1);
+    ParArray1D<int> buffer_start("buffer_start", nneighbor);
 
-  auto &bdvar = vbswarm->bd_var_;
-  auto neighbor_buffer_index = neighbor_buffer_index_;
-  // Loop over active particles and use atomic operations to find indices into buffers if
-  // this particle will be sent.
-  pmb->par_for(
-      PARTHENON_AUTO_LABEL, 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
-        if (swarm_d.IsActive(n)) {
-          bool on_current_mesh_block = true;
-          const int m =
-              swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), on_current_mesh_block);
-          const int bufid = neighbor_buffer_index(m);
-
-          if (m >= 0) {
-            const int bid = Kokkos::atomic_fetch_add(&buffer_counters(m), 1);
-            int buffer_index = bid * particle_size;
-            swarm_d.MarkParticleForRemoval(n);
-            for (int i = 0; i < realPackDim; i++) {
-              bdvar.send[bufid](buffer_index) = vreal(i, n);
-              buffer_index++;
-            }
-            for (int i = 0; i < intPackDim; i++) {
-              bdvar.send[bufid](buffer_index) = static_cast<Real>(vint(i, n));
-              buffer_index++;
+    pmb->par_for(
+        PARTHENON_AUTO_LABEL, 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+          if(swarm_d.IsActive(n)) {
+            bool on_current_mesh_block = true;
+            const int m =
+            swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), on_current_mesh_block);
+            buffer_sorted(n) = SwarmKey(m, n);
+          }
+          else {
+            buffer_sorted(n) = SwarmKey(-1, n);
+          }
+        });
+
+
+    // sort by buffer index
+    sort(buffer_sorted, SwarmKeyComparator(), 0, max_active_index_);
+
+    // use discontinuity check to determine start of a buffer in buffer_sorted array
+    auto &num_particles_to_send = num_particles_to_send_;
+    auto max_active_index = max_active_index_;
+
+    // Zero out number of particles to send before accumulating
+    pmb->par_for(
+        PARTHENON_AUTO_LABEL, 0, NMAX_NEIGHBORS - 1,
+        KOKKOS_LAMBDA(const int n) { num_particles_to_send[n] = 0; });
+
+    pmb->par_for(
+        PARTHENON_AUTO_LABEL, 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
+          auto m = buffer_sorted(n).cell_idx_1d_;
+          // start checks (used for index of particle in buffer)
+          if (m >= 0 && n ==0 ) {
+            buffer_start(m) = 0;
+          }
+          else if (m >= 0 && m != buffer_sorted(n-1).cell_idx_1d_) {
+            buffer_start(m) = n;
+          }
+
+          // end checks (used to to size particle buffers)
+          if (m >= 0 && n == max_active_index ) {
+            num_particles_to_send(m) = n +1;
```

Comment:
@alexrlongne in parthenon we use `clang-format-12` for formatting (I just have this hooked up to vscode so I always format on save) but just mentioning because the CI will fail otherwise.